### ISSUE TITLE
"sql: no rows in result set" error on new HA cluster with mysql as backend

### ISSUE THREAD

**Issue by wych42** [hashicorp/vault#6865]
**Describe the bug**
start fresh vault process with mysql backend and ha_enabled set to true. vault commands returns error: "sql: no rows in result set". such as:
```
vault status
Error checking leader status: Error making API request.

URL: GET http://192.168.12.127:8400/v1/sys/leader
Code: 500. Errors:

* sql: no rows in result set
```
**To Reproduce**
1. start a fresh vault process with config below.
3. run `vault status` 

**Environment:**
* Vault Server Version (retrieve with `vault status`):1.1.3
* Vault CLI Version (retrieve with `vault version`):1.1.3
* Server Operating System/Architecture: Debian amd64

Vault server configuration file(s):

```json
{
    "backend": {
        "mysql": {
            "address": "xxxx:3306",
            "database": "xx",
            "table": "xx",
            "username": "xxx",
            "password": "xxx",
            "ha_enabled": "true"
        }
    },
    "listener": [
        {
            "tcp": {
                "address": "192.168.12.127:8200",
                "tls_disable": 1
            }
        }
    ],
    "api_addr": "http://192.168.12.127:8200",
    "cluster_addr": "http://192.168.12.127:8201",
    "seal": {
        "alicloudkms": {
            "region": "x",
            "access_key": "x",
            "secret_key": "x",
            "kms_key_id": "x"
        }
    },
    "default_lease_ttl": "168h",
    "max_lease_ttl": "720h",
    "ui": true
}
```

**Additional context**
- remove "ha_enabled": "true", vault works as expected, but that would disabled HA.
- set "ha_enabled": 1 will get the same error.

**Comment by wych42** [comment]
This issue seems to be caused by running two vault cluster on same MySQL instance and this line of code: https://github.com/hashicorp/vault/blob/master/physical/mysql/mysql.go#L626

1. MySQL instance M.
2. Vault cluster A is configured with database vault-a on M. Cluster B is configured with database vault-b on M.
3. Start cluster A, the vault process will try and get a lock named `core/lock` in M with timeout of math.MaxInt32, and become leader if it get the lock. When it becomes leader, it will "[put](https://github.com/hashicorp/vault/blob/master/physical/mysql/mysql.go#L611)"(actually sql insert on duplicate update) itself into `vault_lock` table.
4. Start the rest nodes of cluster A, they can't get the named lock `core/lock`, but it's fine because the active node is already running.
5. Start cluster B, it's try and get named lock `core/lock` in M, since the lock is already held by cluster A, this node will wait or skip become leader process, and the cluster-b.vault_lock table is empty. which cause the `no rows in result set` error.

Basically, you can't use one MySQL instance as backend of two different vault instances/clusters.Maybe changing the named lock `core/lock` to `core/lock/{cluster-identifier}` would solve the problem. Or just start a MySQL instance for each vault cluster.

**Comment by noyoded** [comment]
HI 
Have the same issue, how did you find to solve this?

**Comment by wych42** [comment]
@noyoded Don't use one MySQL instance as backend of two different vault instances/clusters.

**Comment by noyoded** [comment]
@wych42, I don't use one instance. 

I have 3 nodes on different vms, and each one has mysql.
BUT, I user Percona XtraDB cluster for HA, and few seconds after starting vault I get the same error for unsealing:

Error checking leader status: Error making API request.

URL: GET http://10.94.12.85:8200/v1/sys/leader
Code: 500. Errors:

* sql: no rows in result set

Thanks ahead! No one managed to solve this so far.

Noy

**Comment by wych42** [comment]
@noyoded I had this issue when I tried to run two vault clusters with same MySQL(instance,node,cluster).

The same MySQL(xx) means: the same MySQL instance if MySQL is running alone or the same MySQL cluster if MySQL is running with Master-Slave mode or HA mode or GTID mode.

The cause for my issue was explained at https://github.com/hashicorp/vault/issues/6865#issuecomment-502095472 ,shortly is the two vault cluster would try to get the same lock in the one MySQL.

If you are setting up two vault clusters like I do, try configure them with two MySQL cluster. Three nodes MySQL with Percona XtraDB cluster is  one instance to valut.

**Comment by jiribaloun** [comment]
Having same issue, on three node Percona XtraDB cluster, running as multi-master fashion.
Has anybody made any progress?

**Comment by SCLogo** [comment]
amny update on this? I really don't want to use separate mysql instance per vault

**Comment by timo-klarshift** [comment]
Was this solved? Why is it closed?

**Comment by jiribaloun** [comment]
> Was this solved? Why is it closed?

I have no clue. Anyway,  I have switched over to raft.
